posts/Swift-Optionals-Are-Monads-Too.md

```diff
@@ -0,0 +1,264 @@
+
+So now that we've added this method to Swift's Optional, it's now an applicative functor!
+
+---
+
+# Monads Ahoy
+
+Last but not least, we come to Monads. Let's again start with a definition in programmer terms.
+
+> A monad is any type that defines flatMap.
+
+Looking at the declaration code for Optional and right away flatMap is screaming in our face. So Swift's Optional type is a Monad without any further modification! Let's make an educated guess how Swift defines flatMap on Optional.
+
+```swift
+func flatMap<U>(@noescape f: (Wrapped) throws -> U?) rethrows -> U? {
+  switch self {
+  case .Some(let x): return try f(x)
+  case .None: return .None
+}
+```
+
+Comparing flatMap and map on Optional you might be asking yourself what's the difference? Really there is only one specific difference and it's that f in map cannot return another Optional whereas f in flatMap can return an Optional.
+
+---
+
+# Why It Matters
+
+So now that we have a good understanding of what each of these abstract concepts mean, it kind of begs the question, what are they good for?
+
+### Mapping out Optionals
+
+Formatting NSDate? can be a bit tricky. Because it's an Optional we are forced into using if/let syntax or heaven forbid force unwrapping the Optional.
+
+```swift
+var date: NSDate?
+var formatted: NSString?
+
+// some time after date *may* have been set or initialized
+
+if let date = date {
+  formatted = NSDateFormatter().stringFromDate(date)
+}
+```
+
+Here's functionally equivalent code using `map`:
+
+```swift
+var date: NSDate?
+
+// some time after date *may* have been set or initialized
+
+var formatted = date.map(NSDateFormatter().stringFromDate)
+```
+
+To be clear, we aren't aiming for code golf. But switching over from the `if/let` to using `map` makes our intent more clear in my opinion. If the date isn't nil, let's format it as a string. If it is nil, our formatted date string is nil. 
+
+Another wonderful usage of `map` is while we are interpolating values in a string:
+
+```swift
+var numberOfFriends: Int?
+
+// some time after numberOfFriends *may* 
+// have been set or initialized
+
+var friendsStatus = numberOfFriends.map { 
+  "You have \($0) friends."
+} ?? "No friends yet." 
+```
+
+This allows us to choose two completely different strings based on whether our `numberOfFriends` variable has a value or not.
+
+One last example of a great usage for `map`. What if we want to read the contents of a url?
```

Comment:
Maybe:

Another example of a great usage...

This may cut down on simple minds thinking this is not the actual last case where it's useful ;-)